### ISSUE TITLE
Overpass query efficiency

### ISSUE THREAD

**Issue by mmd-osm** [EdwardBetts/osm-wikidata#95]
I believe your tool creates queries like in the following example (excerpt) - if not, please ignore the noise :sunglasses: 

```C
[out:xml][bbox:15.739532,73.698950,17.178811,74.700294];
area(3601997163)->.a;
(
    way(area.a)[historic=battlefield][name];
    way(area.a)[historic=castle][name];
    way(area.a)[historic=fort][name];
    way(area.a)[historic=monument][name];
    way(area.a)[historic=museum][name];
    way(area.a)[historic=palace][name];
 );

(._;>;);
out qt;
```

I'd propose to merge query statements for the same key into a single query using regular expression to speed up processing.

```C
[out:xml][bbox:15.739532,73.698950,17.178811,74.700294];
area(3601997163)->.a;
(
    way(area.a)[historic~"^(battlefield|castle|fort|monument|museum|palace)$"][name];
 );

(._;>;);
out qt;
```

In addition, I would extract the `name` filter into a separate query statement for even more significant savings. This turns...

```C
way(area.a)[historic~"^(battlefield|castle|fort|monument|museum|palace)$"][name];
```

into:

```C
   (  way(area.a)[historic~"^(battlefield|castle|fort|monument|museum|palace)$"];
     [... more way queries ...]
   );
    way._[name];
```

full query:
```C
[out:xml][bbox:15.739532,73.698950,17.178811,74.700294];
area(3601997163)->.a;

(
  node(area.a)[historic~"^(battlefield|castle|fort|monument|museum|palace)$"];
  way(area.a)[historic~"^(battlefield|castle|fort|monument|museum|palace)$"];
  rel(area.a)[historic~"^(battlefield|castle|fort|monument|museum|palace)$"];
) -> .b;
(
  node.b[name];
  way.b[name];
  rel.b[name];
);

(._;>;);
out qt;
```

Bottom line: 20x + speedup may be feasible by using an optimized query design.

**Comment by EdwardBetts** [comment]
This is brilliant, I'll try it. Thanks.

**Comment by mmd-osm** [comment]
Looks good, thanks. One short remark with regards to buildings. As some areas have quite a huge number of them, the query could fail because of excessive memory usage. So, that's worth keeping an eye on, and possibly have some special case handling for buildings.

BTW: the rambler.ru instance you're using currently has some issues with updates, since May 31st. I pinged Roland on this (http://wiki.openstreetmap.org/wiki/Overpass_API/status), but he seems to have forgotten to also fix the rambler instance...

**Comment by EdwardBetts** [comment]
Thanks. Maybe for buildings I should include a regex name filter. So I only get buildings with names that I've seen on Wikidata. This would miss a few, but would massively reduce the number of buildings downloaded.

**Comment by EdwardBetts** [comment]
I noticed that some of the replies from rambler.ru were out of date.

The system access Overpass in two different ways. When looking for a match in a specific area it users overpass-api.de, but runs the query in the mapper's browser, then passes the reply back to the server. This means queries are coming from the IP address of the mapper and the rate limiting is less likely to kick in.

The other use of Overpass is for an API to find matches used by this mediawiki gadget: https://www.mediawiki.org/wiki/User:Bjohas/OSMgadget

When requests come in via the API the overpass queries need to be run from the server. It was configured to use overpass-api.de, but this was being the rate limited, so I switched to rambler.ru, I've just tried switching back to overpass-api.de to see if it is still triggering the rate limiting.
